vehicle/mkvi/software/imu/icm20948/icm20948.c

```diff
@@ -0,0 +1,71 @@
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "icm20948.h"
+#include "libs/spi/api.h"
+#include <util/delay.h>
+
+// #include "projects/can_print/can_print.h"
+
+void icm_read_register(uint8_t register_, uint8_t* rx_data) {
+    uint8_t tx_data[2] = { register_ | (1 << 7), 0x0 };
+    uint8_t _rx_data[2];
+    spi_transceive_cs(tx_data, _rx_data, 2);
+    *rx_data = _rx_data[1];
+}
+
+void icm_write_register(uint8_t _register, uint8_t tx_data) {
+    uint8_t _tx_data[2] = { _register, tx_data };
+    spi_transceive_cs(_tx_data, NULL, 2);
+}
+
+// TODO: FIFO use more efficient burst SPI reads
+
+void switch_register_bank(uint8_t bank) {
+    icm_write_register(REG_BANK_SEL, (bank << 4));
+}
+
+void init_magnetometer(void) {
+    switch_register_bank(BANK_3);
+
+    // Close all auxilary I2C buses
+    icm_write_register(I2C_SLV0_CTRL, 0x0);
+    icm_write_register(I2C_SLV1_CTRL, 0x0);
+    icm_write_register(I2C_SLV2_CTRL, 0x0);
+    icm_write_register(I2C_SLV3_CTRL, 0x0);
+
+    icm_write_register(I2C_MST_ODR_CONFIG, 0x04);
+
+    switch_register_bank(BANK_0);
+}
+
+void read_magnetometer(uint8_t register_, uint8_t* rx_data) {
+    switch_register_bank(BANK_3);
+    uint8_t _rx_data;
+
+    // Write the magnetometer addr for reading
+    icm_write_register(I2C_SLV4_ADDR, (I2C_READ << 7) | MAGNETOMETER_ADDR);
```

Comment:
this might be my unfamiliarity with the chip, but couldn't this just be done once in `init_magnetometer`

Reply:
There's some quirks with the magnetometer that I wasn't able to fully resolve. I could read the device ID but no more. I was copying Ian's register writes so I'm not quite sure why this was in every magnetometer read.